# Subject: Classical Data Analysis

## Session 1 - Regression

### Demo 3 -  Linear Regression with Quandl Integration


Implementation of Python StatsModels package with Quandl integration to calculate the Linear regression with one variable.  



# 1 - Quandl Integration

A recurring challenge with any quantitative analysis is the availability of good quality data. Luckily for us, Quandl https://www.quandl.com/ has taken on the data challenge and indexed millions of economics, financial, societal, and country specific datasets. That data is also available through a free API (Application Programming Interface) supported by the Quandl Python package (https://www.quandl.com/tools/python).

# 2 - Register in Quandl

Get your Username, Password and Access key. Explore the Quandl website.

# 3 - Installation

For instructions on installing Quandl for Python check out PyPi (https://docs.python.org/2/install/) or the Github page (https://github.com/quandl/quandl-python).

The installation process varies depending on your python version and system used. However in most cases the following should work. Execute the following Pip command from CMD (Windows):

### pip install quandl

Alternatively on some systems python3 may use a different pip executable and may need to be installed via an alternate pip command. For example:

### pip3 install quandl

# 4 - Retrieving Data

There are two methods for retrieving data in Python: the Quick method and the Detailed method. The latter is more suitable to application programming. Both methods work with Quandl's two types of data structures: time-series (dataset) data and non-time series (datatable).

Retrieving data can be achieved easily using the two methods quandl.get for datasets and quandl.get_table for datatables. In both cases we strongly recommend that you set your api key via:

In [1]:
import quandl
quandl.ApiConfig.api_key = 'wagAy5tFsmUZ84CH3Ng8' # A valid API key is required to retrieve data. Please check your API key and try again. You can find your API key under your account settings.

## Datasets (Time-Series)

The following quick call can be used to retrieve a dataset:

In [2]:
import quandl
data1 = quandl.get('NSE/OIL')

This example finds all data points for the dataset NSE/OIL and stores (https://www.quandl.com/data/NSE/OIL) them in a pandas dataframe. You can then view the dataframe with data.head().

In [3]:
data1.head()

Open    High     Low     Last    Close  Total Trade Quantity  \
Date                                                                         
2009-09-30  1096.0  1156.7  1090.0  1135.00  1141.20            19748012.0   
2009-10-01  1102.0  1173.7  1102.0  1167.00  1166.35             3074254.0   
2009-10-05  1152.0  1165.9  1136.6  1143.00  1140.55              919832.0   
2009-10-06  1149.8  1157.2  1132.1  1143.30  1144.90              627957.0   
2009-10-07  1153.8  1160.7  1140.0  1141.45  1141.60              698216.0   

            Turnover (Lacs)  
Date                         
2009-09-30        223877.07  
2009-10-01         35463.78  
2009-10-05         10581.13  
2009-10-06          7185.90  
2009-10-07          8032.98

Loading CSV data in Python with pandas.

Use this API call to get data in CSV format: https://www.quandl.com/api/v3/datasets/NSE/OIL.csv?api_key=wagAy5tFsmUZ84CH3Ng8

In [4]:
import pandas as pd
data12=pd.read_csv("https://www.quandl.com/api/v3/datasets/NSE/OIL.csv?api_key=wagAy5tFsmUZ84CH3Ng8")
data12.head()

Date    Open    High     Low    Last   Close  Total Trade Quantity  \
0  2017-10-03  353.00  356.90  348.85  354.50  354.75              689435.0   
1  2017-09-29  345.95  354.95  340.10  353.65  353.65              861320.0   
2  2017-09-28  349.90  354.95  335.80  344.70  348.30             2079490.0   
3  2017-09-27  349.70  356.10  345.10  350.00  350.55             1493047.0   
4  2017-09-26  334.00  350.00  332.60  349.70  349.35             1911979.0   

   Turnover (Lacs)  
0          2432.93  
1          3013.87  
2          7185.88  
3          5236.91  
4          6593.90

Loading CSV data in Python with pandas using a file exported from Quandl.

In [5]:
import pandas as pd
data13=pd.read_csv("C:/Users/francisco.sacramento/Desktop/Master_Big_Data_Phyton/6_Exercices/Classical Data Analysis/NSE-OIL.csv")
data13.head()

Date   Open    High     Low   Last   Close  Total Trade Quantity  \
0  2017-09-27  349.7  356.10  345.10  350.0  350.55             1493047.0   
1  2017-09-26  334.0  350.00  332.60  349.7  349.35             1911979.0   
2  2017-09-25  325.3  333.75  321.20  332.0  332.50              995497.0   
3  2017-09-22  327.9  332.65  321.50  328.9  328.70             1428072.0   
4  2017-09-21  328.0  330.20  322.05  326.4  328.30              769453.0   

   Turnover (Lacs)  
0          5236.91  
1          6593.90  
2          3278.67  
3          4686.54  
4          2522.00

However in some cases is recommended applying filters to streamline the results. To do this you may want to specify some additional filters.

In [6]:
import quandl
data2 = quandl.get('NSE/OIL', start_date='2010-01-01', end_date='2014-01-01')

This revised query will find all data points annually for the dataset NSE/OIL between the year 2010 and 2014. 

In [7]:
data2.head()

Open    High      Low    Last    Close  Total Trade Quantity  \
Date                                                                          
2010-01-04  1259.90  1259.9  1235.00  1242.5  1240.30               41769.0   
2010-01-05  1259.55  1265.0  1237.05  1249.9  1250.55              129799.0   
2010-01-06  1255.10  1268.0  1242.50  1250.0  1248.80               63129.0   
2010-01-07  1246.10  1255.0  1231.30  1241.5  1241.40               36209.0   
2010-01-08  1248.90  1268.0  1230.00  1265.0  1262.95               64337.0   

            Turnover (Lacs)  
Date                         
2010-01-04           519.53  
2010-01-05          1616.00  
2010-01-06           790.45  
2010-01-07           451.02  
2010-01-08           807.72

## Datatables

Datatables work similarly to datasets but provide more flexibility when it comes to filtering. For example a simple way to retrieve datatable information would be:

In [8]:
import quandl
data3 = quandl.get_table('ZACKS/FC')

In [9]:
data3.head()

m_ticker ticker  comp_name comp_name_2 exchange currency_code  \
None                                                                 
0        AAPL   AAPL  APPLE INC  Apple Inc.     NSDQ           USD   
1        AAPL   AAPL  APPLE INC  Apple Inc.     NSDQ           USD   
2        AAPL   AAPL  APPLE INC  Apple Inc.     NSDQ           USD   
3        AAPL   AAPL  APPLE INC  Apple Inc.     NSDQ           USD   
4        AAPL   AAPL  APPLE INC  Apple Inc.     NSDQ           USD   

     per_end_date per_type per_code  per_fisc_year       ...         \
None                                                     ...          
0      2011-09-30        A     None           2011       ...          
1      2012-09-30        A     None           2012       ...          
2      2013-09-30        A     None           2013       ...          
3      2014-09-30        A     None           2014       ...          
4      2015-09-30        A     None           2015       ...          

      stock_based_compsn_qd  cash_flow_oper_activity_qd  \
None                                                      
0                       NaN                         NaN   
1                       NaN                         NaN   
2                       NaN                         NaN   
3                       NaN                         NaN   
4                       NaN                         NaN   

      net_change_prop_plant_equip_qd comm_stock_div_paid_qd  \
None                                                          
0                                NaN                    NaN   
1                                NaN                    NaN   
2                                NaN                    NaN   
3                                NaN                    NaN   
4                                NaN                    NaN   

     pref_stock_div_paid_qd  tot_comm_pref_stock_div_qd  wavg_shares_out  \
None                                                                       
0                      None                         NaN         6469.806   
1                      None                         NaN         6543.726   
2                      None                         NaN         6477.320   
3                      None                         NaN         6085.572   
4                      None                         NaN         5753.421   

      wavg_shares_out_diluted  eps_basic_net  eps_diluted_net  
None                                                           
0                    6556.515         4.0071           3.9543  
1                    6617.483         6.3800           6.3100  
2                    6521.634         5.7200           5.6800  
3                    6122.663         6.4900           6.4500  
4                    5793.069         9.2800           9.2200  

[5 rows x 249 columns]

This is the syntax for calling an entire table. Given the volume of data stored in datatables, this call will retrieve the first page of the ZACKS/FC datatable. You may turn on pagination to return more data by using:

In [10]:
import quandl
data4 = quandl.get_table('ZACKS/FC', paginate=True)

In [11]:
data4

m_ticker ticker        comp_name              comp_name_2 exchange  \
None                                                                      
0        AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
1        AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
2        AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
3        AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
4        AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
5        AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
6        AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
7        AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
8        AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
9        AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
10       AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
11       AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
12       AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
13       AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
14       AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
15       AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
16       AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
17       AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
18       AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
19       AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
20       AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
21       AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
22       AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
23       AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
24       AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
25       AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
26       AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
27       AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
28       AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
29       AAPL   AAPL        APPLE INC               Apple Inc.     NSDQ   
...       ...    ...              ...                      ...      ...   
940         J    XOM  EXXON MOBIL CRP  Exxon Mobil Corporation     NYSE   
941         J    XOM  EXXON MOBIL CRP  Exxon Mobil Corporation     NYSE   
942         J    XOM  EXXON MOBIL CRP  Exxon Mobil Corporation     NYSE   
943         J    XOM  EXXON MOBIL CRP  Exxon Mobil Corporation     NYSE   
944         J    XOM  EXXON MOBIL CRP  Exxon Mobil Corporation     NYSE   
945         J    XOM  EXXON MOBIL CRP  Exxon Mobil Corporation     NYSE   
946         J    XOM  EXXON MOBIL CRP  Exxon Mobil Corporation     NYSE   
947         J    XOM  EXXON MOBIL CRP  Exxon Mobil Corporation     NYSE   
948         J    XOM  EXXON MOBIL CRP  Exxon Mobil Corporation     NYSE   
949         J    XOM  EXXON MOBIL CRP  Exxon Mobil Corporation     NYSE   
950         J    XOM  EXXON MOBIL CRP  Exxon Mobil Corporation     NYSE   
951         J    XOM  EXXON MOBIL CRP  Exxon Mobil Corporation     NYSE   
952         J    XOM  EXXON MOBIL CRP  Exxon Mobil Corporation     NYSE   
953         J    XOM  EXXON MOBIL CRP  Exxon Mobil Corporation     NYSE   
954         J    XOM  EXXON MOBIL CRP  Exxon Mobil Corporation     NYSE   
955         J    XOM  EXXON MOBIL CRP  Exxon Mobil Corporation     NYSE   
956         J    XOM  EXXON MOBIL CRP  Exxon Mobil Corporation     NYSE   
957         J    XOM  EXXON MOBIL CRP  Exxon Mobil Corporation     NYSE   
958         J    XOM  EXXON MOBIL CRP  Exxon Mobil Corporation     NYSE   
959         J    XOM  EXXON MOBIL CRP  Exxon Mobil Corporation     NYSE   
960         J    XOM  EXXON MO

This will retrieve multiple pages of data and merge them together as if they were one large page. In some cases, however, you will still exceed the request limit. In this case we recommend you filter your data using the available query parameters, as in the following example:

In [12]:
import quandl
data5 = quandl.get_table('ZACKS/FC', paginate=True, ticker=['AAPL', 'MSFT'], per_end_date={'gte': '2015-01-01'}, qopts={'columns':['ticker', 'per_end_date']})

In [13]:
data5

ticker per_end_date
None                    
0      AAPL   2015-09-30
1      AAPL   2016-09-30
2      AAPL   2015-03-31
3      AAPL   2015-06-30
4      AAPL   2015-09-30
5      AAPL   2015-12-31
6      AAPL   2016-03-31
7      AAPL   2016-06-30
8      AAPL   2016-09-30
9      AAPL   2016-12-31
10     AAPL   2017-03-31
11     AAPL   2017-06-30
12     MSFT   2015-06-30
13     MSFT   2016-06-30
14     MSFT   2017-06-30
15     MSFT   2015-03-31
16     MSFT   2015-06-30
17     MSFT   2015-09-30
18     MSFT   2015-12-31
19     MSFT   2016-03-31
20     MSFT   2016-06-30
21     MSFT   2016-09-30
22     MSFT   2016-12-31
23     MSFT   2017-03-31
24     MSFT   2017-06-30

In this query we are asking for more pages of data, ticker values of either AAPL or MSFT and a per_end_date that is greater than or equal to 2015-01-01. We are also filtering the returned columns on ticker, per_end_date and comp_name rather than all available columns. The output format is pandas.

# Load Quandl Data Directly Into Python

All datasets and databases on Quandl are available directly in Python, using the Quandl Python module.

For example: https://www.quandl.com/data/UTOR/EXTUR_ESP-Outbound-Tourism-Spain

Get from Export Data option - Phyton librarySign in Quandl. You must sign in.

In [14]:
data6 = quandl.get("UTOR/EXTUR_ESP", authtoken="wagAy5tFsmUZ84CH3Ng8")

In [15]:
data6.head()

Departures - Thousands  \
Date                                 
1995-12-31                  3648.0   
1996-12-31                  4224.0   
1997-12-31                  3980.0   
1998-12-31                  4794.0   
1999-12-31                  3519.0   

            Tourism expenditure in other countries - US$ Mn  Travel - US$ Mn  \
Date                                                                           
1995-12-31                                           5826.0           4539.0   
1996-12-31                                           6332.0           5057.0   
1997-12-31                                           6163.0           4681.0   
1998-12-31                                           7119.0           5288.0   
1999-12-31                                           7803.0           5862.0   

            Passenger transport - US$ Mn  
Date                                      
1995-12-31                        1287.0  
1996-12-31                        1275.0  
1997-12-31                        1482.0  
1998-12-31                        1831.0  
1999-12-31                        1941.0

# 5 - Regression model with Statsmodels and without a constant

In [16]:
import statsmodels.api as sm

X = data6["Departures - Thousands"]
y = data6["Travel - US$ Mn"]

Note the difference in argument order:

In [17]:
model = sm.OLS(y, X).fit() ## sm.OLS(output, input)


Make the predictions by the model:

In [18]:
predictions = model.predict(X) 

Print out the statistics:

In [19]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        Travel - US$ Mn   R-squared:                       0.976
Model:                            OLS   Adj. R-squared:                  0.975
Method:                 Least Squares   F-statistic:                     784.6
Date:                Tue, 03 Oct 2017   Prob (F-statistic):           6.48e-17
Time:                        19:53:37   Log-Likelihood:                -180.83
No. Observations:                  20   AIC:                             363.7
Df Residuals:                      19   BIC:                             364.7
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==========================================================================================
                             coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------------------
Departures - Thousands     1.4936      0.053     28.011      0.000         1.382     1.605
==============================================================================
Omnibus:                        1.502   Durbin-Watson:                   0.912
Prob(Omnibus):                  0.472   Jarque-Bera (JB):                1.247
Skew:                           0.560   Prob(JB):                        0.536
Kurtosis:                       2.510   Cond. No.                         1.00
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Interpreting the Table 


The coefficient of 1.4936 means that as the Departures - Thousands variable increases by 1, the predicted value of Travel - US$ Mn increases by 1.4936. A few other important values are the R-squared of 0.976 — the percentage of variance our model explains.

# 6 - Regression model with Statsmodels and with a constant

In [20]:
import statsmodels.api as sm

X = data6["Departures - Thousands"]
y = data6["Travel - US$ Mn"]

In [21]:
X = sm.add_constant(X) 

In [22]:
model = sm.OLS(y, X).fit() ## sm.OLS(output, input)

In [23]:
predictions = model.predict(X)

In [24]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        Travel - US$ Mn   R-squared:                       0.877
Model:                            OLS   Adj. R-squared:                  0.870
Method:                 Least Squares   F-statistic:                     128.4
Date:                Tue, 03 Oct 2017   Prob (F-statistic):           1.26e-09
Time:                        19:53:45   Log-Likelihood:                -180.42
No. Observations:                  20   AIC:                             364.8
Df Residuals:                      18   BIC:                             366.8
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==========================================================================================
                             coef    std err          t      P>|t|      [95.0% Conf. Int.]
------------------------------------------------------------------------------------------
const                    939.2279   1084.358      0.866      0.398     -1338.923  3217.379
Departures - Thousands     1.3975      0.123     11.332      0.000         1.138     1.657
==============================================================================
Omnibus:                        2.126   Durbin-Watson:                   0.871
Prob(Omnibus):                  0.345   Jarque-Bera (JB):                1.741
Skew:                           0.673   Prob(JB):                        0.419
Kurtosis:                       2.475   Cond. No.                     2.02e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.02e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### Interpreting the Table 
With the constant term the coefficients are different. Without a constant we are forcing our model to go through the origin, but now we have a y-intercept at 939.2279. We also changed the slope of the Departures - Thousands predictor from 1.4936 to 1.3975.
A few other important values are modification of the R-squared, from 0.976 to 0.877.